utils/ssh_client.py

```
@@ -0,0 +1,133 @@
+        if path is None:
+            self.client.load_host_keys(os.path.expanduser('~/.ssh/known_hosts'))
+        else:
+            self.client.load_host_keys(path)
+
+    def clear_host_keys(self):
+        self.client.get_host_keys().clear()
+
+    def _connect_with_password(self, **kwargs):
+        if self.password is None:
+            message = 'Password is not specified for {}@{}'.format(self.username, self.host)
+            logger.error(message)
+            raise SSHConnectionFailed(message)
+        self.client.connect(self.host, username=self.username, password=self.password, **kwargs)
+
+    def _connect_with_key(self, **kwargs):
+        if self.key is None:
+            message = 'Private key is not specified for {}@{}. Attempt to connect using credentials'.format(
+                self.username, self.host)
+            logger.warning(message)
+            raise SSHKeyMissing(message)
+        if self.passphrase is None:
+            message = 'Passphrase is not specified for {}'.format(self.key)
+            logger.error(message)
+            raise SSHConnectionFailed(message)
+        self.client.connect(self.host, username=self.username, key_filename=self.key, passphrase=self.passphrase,
+                            **kwargs)
+
+    def connect(self, use_credentials=False, **kwargs):
+        if self.allow_unknown_hosts:
+            self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+        if self.username is None:
+            message = 'Username is not specified for {}'.format(self.host)
+            logger.error(message)
+            raise SSHConnectionFailed(message)
+
+        try:
+            if use_credentials:
+                self._connect_with_password(**kwargs)
+            else:
+                try:
+                    self._connect_with_key(**kwargs)
+                except SSHKeyMissing:
+                    self._connect_with_password(**kwargs)
+
+        except paramiko.SSHException:
```

You are hiding error here, it's very bad.

---

Done